File: solutions/005/Peter/main.py
def cons(a, b):
    def pair(f):
        return f(a, b)
    return pair

def car(pair):
    def returnfirst(a,b):
        return a
    return pair(returnfirst)

def cdr(pair):
    def returnlast(a,b):
        return b
    return pair(returnlast)

def add(a,b):
    return a+b

# print(cons(1,2).__closure__[0].cell_contents)
something = cons(1,2)
something(print)
print(something(add))
print(cdr(something))
print(car(something))

File: solutions/005/Peter/main.cpp
#include <iostream>
#include <utility>

using namespace std;

pair<int, int> cons(int a, int b) {
    auto f = [&](int a, int b){return make_pair(a, b);};
    auto pair = [&]() { return f(a, b); };
    return pair();
}

int car(pair<int, int> pair){
    auto returnfirst = [&](){return pair.first;};
    return returnfirst();
}

int cdr(pair<int, int> pair){
    auto returnlast = [&](){return pair.second;};
    return returnlast();
}

int main() {
    auto pair = cons(3, 4);
    cout << car(pair) << endl << cdr(pair) << endl;
    return 0;
}